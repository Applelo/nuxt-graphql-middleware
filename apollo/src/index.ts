import { ApolloServer } from '@apollo/server'
import { startStandaloneServer } from '@apollo/server/standalone'
import data from './data.json' assert { type: 'json' }

let users = [...data]
let idIncrement = users.length

function getId() {
  idIncrement++
  return idIncrement
}

const typeDefs = `#graphql
  type User {
    id: Int!
    firstName: String!
    lastName: String!
    email: String!
    description: String
    dateOfBirth: String
  }

  input UserData {
    firstName: String!
    lastName: String!
    email: String!
    description: String
    dateOfBirth: String
  }

  type Query {
    users: [User!]!
    userById(id: ID!): User
  }

  type Mutation {
    createUser(user: UserData!): User!
    deleteUser(id: Int!): Boolean
  }
`

const resolvers = {
  Query: {
    users: () => {
      return users
    },
    userById: (_: any, args: any) => {
      const id = parseInt(args.id)
      return users.find((v) => v.id === id)
    },
  },
  Mutation: {
    createUser: (_: any, args: any) => {
      const user = { id: getId(), ...args.user }
      users.push(user)
      return user
    },
    deleteUser: (_: any, args: any) => {
      users = users.filter((v) => v.id !== args.id)
      return true
    },
  },
}

const server = new ApolloServer({
  typeDefs,
  resolvers,
})

const { url } = await startStandaloneServer(server, { listen: { port: 4000 } })

console.log(`GraphQL server listening at: ${url}`)
