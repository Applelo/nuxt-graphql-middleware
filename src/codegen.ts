import { generate } from '@graphql-codegen/cli'
import * as PluginTypescript from '@graphql-codegen/typescript'
import * as PluginTypescriptOperations from '@graphql-codegen/typescript-operations'
import * as PluginNuxtGraphqlMiddleware from './codegen/plugin'
import * as PluginNuxtGraphqlMiddlewareDocuments from './codegen/pluginDocuments'
import * as PluginSchemaAst from '@graphql-codegen/schema-ast'
import { TypeScriptDocumentsPluginConfig } from '@graphql-codegen/typescript-operations'
import { GraphqlMiddlewareTemplate } from './types'

function pluginLoader(name: string): Promise<any> {
  if (name === '@graphql-codegen/typescript') {
    return Promise.resolve(PluginTypescript)
  } else if (name === '@graphql-codegen/typescript-operations') {
    return Promise.resolve(PluginTypescriptOperations)
  } else if (name === '@graphql-codegen/typescript-nuxt-graphql-middleware') {
    return Promise.resolve(PluginNuxtGraphqlMiddleware)
  } else if (
    name === '@graphql-codegen/typescript-nuxt-graphql-middleware-documents'
  ) {
    return Promise.resolve(PluginNuxtGraphqlMiddlewareDocuments)
  } else {
    return Promise.resolve(PluginSchemaAst)
  }
}

export interface CodegenResult {
  filename: string
  content: string
}
export function generateSchema(
  url: string,
  dest: string,
): Promise<CodegenResult[]> {
  return generate(
    {
      schema: url,
      pluginLoader,
      generates: {
        [dest]: {
          plugins: ['schema-ast'],
          config: {
            sort: true,
          },
        },
      },
    },
    true,
  )
}

export function generateTemplates(
  documents: string[],
  schemaPath: string,
  config: TypeScriptDocumentsPluginConfig,
): Promise<CodegenResult[]> {
  return generate(
    {
      schema: schemaPath,
      pluginLoader,
      silent: true,
      documents,
      generates: {
        [GraphqlMiddlewareTemplate.OperationTypes]: {
          plugins: ['typescript', 'typescript-operations'],
          config,
        },
        [GraphqlMiddlewareTemplate.ComposableContext]: {
          plugins: ['typescript-nuxt-graphql-middleware'],
        },
        [GraphqlMiddlewareTemplate.Documents]: {
          plugins: ['typescript-nuxt-graphql-middleware-documents'],
        },
      },
    },
    false,
  )
}
