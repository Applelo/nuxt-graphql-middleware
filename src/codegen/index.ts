import { generate, executeCodegen } from '@graphql-codegen/cli'
import * as PluginTypescript from '@graphql-codegen/typescript'
import * as PluginTypescriptOperations from '@graphql-codegen/typescript-operations'
import * as PluginSchemaAst from '@graphql-codegen/schema-ast'
import * as PluginNuxtGraphqlMiddleware from './plugin'
import * as PluginNuxtGraphqlMiddlewareDocuments from './pluginDocuments'
import { GraphqlMiddlewareTemplate } from './../runtime/settings'
import { ModuleOptions } from './../module'

function pluginLoader(name: string): Promise<any> {
  if (name === '@graphql-codegen/typescript') {
    return Promise.resolve(PluginTypescript)
  } else if (name === '@graphql-codegen/typescript-operations') {
    return Promise.resolve(PluginTypescriptOperations)
  } else if (name === '@graphql-codegen/typescript-nuxt-graphql-middleware') {
    return Promise.resolve(PluginNuxtGraphqlMiddleware)
  } else if (
    name === '@graphql-codegen/typescript-nuxt-graphql-middleware-documents'
  ) {
    return Promise.resolve(PluginNuxtGraphqlMiddlewareDocuments)
  } else {
    return Promise.resolve(PluginSchemaAst)
  }
}

export interface CodegenResult {
  filename: string
  content: string
}
export function generateSchema(
  url: string,
  dest: string,
  writeToDisk: boolean,
): Promise<CodegenResult> {
  return generate(
    {
      schema: url,
      pluginLoader,
      silent: true,
      errorsOnly: true,
      generates: {
        [dest]: {
          plugins: ['schema-ast'],
          config: {
            sort: true,
          },
        },
      },
    },
    writeToDisk,
  ).then((v) => v[0])
}

export function generateTemplates(
  documents: string[],
  schemaPath: string,
  options: ModuleOptions,
): Promise<CodegenResult[]> {
  return executeCodegen({
    schema: schemaPath,
    pluginLoader,
    silent: true,
    errorsOnly: true,
    documents,
    generates: {
      [GraphqlMiddlewareTemplate.OperationTypes]: {
        plugins: ['typescript', 'typescript-operations'],
        config: options.codegenConfig,
      },
      [GraphqlMiddlewareTemplate.ComposableContext]: {
        plugins: ['typescript-nuxt-graphql-middleware'],
        config: {
          serverApiPrefix: options.serverApiPrefix,
        },
      },
      [GraphqlMiddlewareTemplate.Documents]: {
        plugins: ['typescript-nuxt-graphql-middleware-documents'],
      },
    },
  })
}
