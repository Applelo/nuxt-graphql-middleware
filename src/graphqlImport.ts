const fragmentImport = require('@graphql-fragment-import/lib/inline-imports')

export default function (path: string, resolver: any) {
  return fragmentImport(path, {
    resolveImport(identifier: string) {
      return resolver(identifier)
    },
    resolveOptions: {
      basedir: './',
    },
  })
}
