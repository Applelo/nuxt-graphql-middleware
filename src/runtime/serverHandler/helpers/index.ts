import { QueryObject } from 'ufo'
import type { H3Event } from 'h3'
import { createError } from 'h3'
import type { FetchOptions } from 'ofetch'
import type { GraphqlMiddlewareConfig } from './../../../types'
import { GraphqlMiddlewareOperation } from './../../../types'

// Get the variables from query parameters.
//
// For simple cases, e.g. when only primitive variable values are sent:
// ?path=foobar
//
// In complex cases, the entire variables are sent as a JSON encoded string:
// ?__variables=%7B%22foobar%22:%7B%22path%22:%22%22%7D%7D
export function queryParamToVariables(query: QueryObject) {
  try {
    if (query.__variables && typeof query.__variables === 'string') {
      return JSON.parse(query.__variables)
    }
  } catch (_e) {}

  return query
}

/**
 * Get the URL of the GraphQL endpoint.
 */
export function getEndpoint(
  moduleConfig: GraphqlMiddlewareConfig,
  event: H3Event,
  operation: GraphqlMiddlewareOperation,
  operationName: string,
): string {
  if (typeof moduleConfig.graphqlEndpoint === 'string') {
    return moduleConfig.graphqlEndpoint
  } else if (typeof moduleConfig.graphqlEndpoint === 'function') {
    const endpoint = moduleConfig.graphqlEndpoint(
      event,
      operation,
      operationName,
    )
    if (endpoint && typeof endpoint === 'string') {
      return endpoint
    }
  }
  throw new Error('Failed to determine endpoint for GraphQL server.')
}

/**
 * Get the options for the $fetch request to the GraphQL server.
 */
export function getFetchOptions(
  moduleConfig: GraphqlMiddlewareConfig,
  event: H3Event,
  operation: GraphqlMiddlewareOperation,
  operationName: string,
): FetchOptions {
  if (typeof moduleConfig.serverFetchOptions === 'function') {
    return (
      moduleConfig.serverFetchOptions(event, operation, operationName) || {}
    )
  } else if (typeof moduleConfig.serverFetchOptions === 'object') {
    return moduleConfig.serverFetchOptions
  }

  return {}
}

function throwError(statusMessage: string, statusCode = 400): never {
  throw createError({
    statusCode,
    statusMessage,
  })
}

/**
 * Assure that the request is valid.
 */
export function validateRequest(
  method?: string,
  operation?: GraphqlMiddlewareOperation,
  name?: string,
  documents?: Record<string, Record<string, string>>,
): void {
  if (method !== 'POST' && method !== 'GET') {
    throwError('Method not allowed.', 405)
  }

  if (
    operation !== GraphqlMiddlewareOperation.Query &&
    operation !== GraphqlMiddlewareOperation.Mutation
  ) {
    throwError('Unknown operation.')
  }

  // Only allow POST for /mutation/:name.
  if (method === 'POST' && operation !== GraphqlMiddlewareOperation.Mutation) {
    throwError('Queries must be a GET request.')
  }

  // Only allow GET for /query/:name.
  if (method === 'GET' && operation !== GraphqlMiddlewareOperation.Query) {
    throwError('Mutations must be a POST request.')
  }

  if (!name) {
    throwError('Missing name for operation.')
  }

  if (!documents) {
    throwError('Failed to load GraphQL documents', 500)
  }

  if (!documents[operation][name]) {
    throwError(`Operation "${operation}" with name "${name}" not found.`)
  }
}
