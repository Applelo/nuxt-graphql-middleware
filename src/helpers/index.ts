import { existsSync, promises as fsp } from 'node:fs'
import { resolveFiles, resolveAlias, useLogger } from '@nuxt/kit'
import { resolve } from 'pathe'
import type { Resolver } from '@nuxt/kit'
// @ts-ignore
import fragmentImport from '@graphql-fragment-import/lib/inline-imports.js'
import { validateGraphQlDocuments } from '@graphql-tools/utils'
import { loadSchema } from '@graphql-tools/load'
import Table from 'cli-table'
import chalk from 'chalk'
import type {
  GraphQLSchema,
  GraphQLError,
  DocumentNode,
  OperationDefinitionNode,
} from 'graphql'
import { parse, Source } from 'graphql'
import { falsy } from '../runtime/helpers'
import { CodegenResult, generateSchema, generateTemplates } from './../codegen'
import { GraphqlMiddlewareDocument } from './../types'
import { ModuleOptions } from './../module'

export const logger = useLogger('nuxt-graphql-middleware')

export const defaultOptions: ModuleOptions = {
  codegenConfig: {
    exportFragmentSpreadSubTypes: true,
    preResolveTypes: true,
    skipTypeNameForRoot: true,
    skipTypename: true,
    useTypeImports: true,
    onlyOperationTypes: true,
    namingConvention: {
      enumValues: 'change-case-all#upperCaseFirst',
    },
  },
  downloadSchema: true,
  schemaPath: './schema.graphql',
  serverApiPrefix: '/api/graphql_middleware',
  graphqlEndpoint: '',
  debug: false,
  includeComposables: true,
  documents: [],
  autoImportPatterns: ['**/*.{gql,graphql}', '!node_modules'],
}

/**
 * Import and inline fragments in GraphQL documents.
 */
export function inlineFragments(source: string, resolver: any): string {
  return fragmentImport(source, {
    resolveImport(identifier: string) {
      return resolver(identifier)
    },
    resolveOptions: {
      basedir: './',
    },
  })
}

function validateDeprecated(options: any) {
  const deprecatedKeys = [
    'graphqlEndpoint',
    'serverFetchOptions',
    'onServerResponse',
    'onServerError',
  ]

  deprecatedKeys.forEach((key) => {
    if (typeof options[key] === 'function') {
      logger.error(
        `Providing a function for "${key}" via nuxt.config.ts has been removed. Please move the configuration to ~/app/graphqlMiddleware.serverOptions.ts.`,
      )

      if (key === 'graphqlEndpoint') {
        logger.info(`
import { defineGraphqlServerOptions } from '#graphql-server-options'
import { getHeader } from 'h3'
import acceptLanguageParser from 'accept-language-parser';

export default defineGraphqlServerOptions({
  graphqlEndpoint(event, operation, operationName) {
    // Get accepted languages.
    const acceptLanguage = getHeader('accept-language')
    const languages = acceptLanguageParser.parse(acceptLanguage);

    // Use first match or fallback to English.
    const language = languages[0]?.code || 'en'
    return \`https://api.example.com/\${language}/graphql\`
  }
})`)
      }

      if (key === 'serverFetchOptions') {
        logger.info(`
import { defineGraphqlServerOptions } from '#graphql-server-options'
import { getHeader } from 'h3'

// Pass the cookie from the client request to the GraphQL request.
export default defineGraphqlServerOptions({
  serverFetchOptions(event, operation, operationName) {
    return {
      headers: {
        Cookie: getHeader(event, 'cookie')
      }
    }
  }
})`)
      }

      if (key === 'onServerResponse') {
        logger.info(`
import { defineGraphqlServerOptions } from '#graphql-server-options'
import type { H3Event } from 'h3'
import type { FetchResponse } from 'ofetch'

export default defineGraphqlServerOptions({
  onServerResponse(event, graphqlResponse) {
    // Set a static header.
    event.node.res.setHeader('x-nuxt-custom-header', 'A custom header value')

    // Pass the set-cookie header from the GraphQL response to the client.
    const setCookie = graphqlResponse.headers.get('set-cookie')
    if (setCookie) {
      event.node.res.setHeader('set-cookie', setCookie)
    }

    // Add additional properties to the response.
    graphqlResponse._data.__customProperty = ['My', 'values']

    // Return the GraphQL response.
    return graphqlResponse._data
  }
})`)
      }

      if (key === 'onServerError') {
        logger.info(`
import { defineGraphqlServerOptions } from '#graphql-server-options'
import type { H3Event } from 'h3'
import type { FetchError } from 'ofetch'

export default defineGraphqlServerOptions({
  onServerError( event, error, operation, operationName) {
    event.setHeader('cache-control', 'no-cache')
    return {
      data: {},
      errors: [error.message]
    }
  }
})`)
      }

      throw new TypeError('Invalid configuration for graphqlMiddleware.' + key)
    }
  })
}

/**
 * Validate the module options.
 */
export function validateOptions(options: Partial<ModuleOptions>) {
  if (!options.graphqlEndpoint) {
    throw new Error('Missing graphqlEndpoint.')
  }

  validateDeprecated(options)
}

/**
 * Get the path to the GraphQL schema.
 */
export async function getSchemaPath(
  options: ModuleOptions,
  resolver: Resolver['resolve'],
  writeToDisk = false,
): Promise<string> {
  const dest = resolver(options.schemaPath!)
  if (!options.downloadSchema) {
    const fileExists = await fsp
      .access(dest)
      .then(() => true)
      .catch(() => false)
    if (!fileExists) {
      logger.error(
        '"downloadSchema" is set to false but no schema exists at ' + dest,
      )
      throw new Error('Missing GraphQL schema.')
    }
    return dest
  }
  if (!options.graphqlEndpoint) {
    throw new Error('Missing graphqlEndpoint config.')
  }
  await generateSchema(options, dest, writeToDisk)
  return dest
}

/**
 * Read the GraphQL files for the given patterns.
 */
export async function autoImportDocuments(
  patterns: string[] = [],
  srcResolver: Resolver['resolve'],
): Promise<GraphqlMiddlewareDocument[]> {
  if (!patterns.length) {
    return Promise.resolve([])
  }
  const files = (
    await resolveFiles(srcResolver(), patterns, {
      followSymbolicLinks: false,
    })
  ).filter((path) => {
    return !path.includes('schema.gql') && !path.includes('schema.graphql')
  })

  return Promise.all(
    files.map((filename) => {
      return fsp.readFile(filename).then((v) => {
        const content = v.toString().trim()
        return {
          content,
          filename,
        }
      })
    }),
  )
}

export function buildDocuments(
  providedDocuments: string[] = [],
  autoImportPatterns: string[],
  resolver: Resolver['resolve'],
): Promise<GraphqlMiddlewareDocument[]> {
  return autoImportDocuments(autoImportPatterns, resolver)
    .then((importedDocuments) => {
      return [
        ...importedDocuments,
        ...providedDocuments.map((content) => {
          return {
            content,
            filename: 'nuxt.config.ts',
          }
        }),
      ].filter((v) => v.content)
    })
    .then((documents) => {
      return documents
        .map((v) => {
          try {
            return {
              content: inlineFragments(v.content, resolveAlias),
              filename: v.filename,
            }
          } catch (e) {
            logger.error(
              'Failed to inline fragments for document: ' + v.filename,
            )
          }
          return null
        })
        .filter(falsy)
    })
}

export function parseDocument(
  document: GraphqlMiddlewareDocument,
  srcDir: string,
): DocumentNode {
  let name = document.filename ? document.filename.replace(srcDir, '') : ''
  if (name.charAt(0) === '/') {
    name = name.slice(1)
  }
  const source = new Source(document.content, name)
  return parse(source)
}

export function validateDocuments(
  schema: GraphQLSchema,
  documents: GraphqlMiddlewareDocument[],
  srcDir: string,
): GraphqlMiddlewareDocument[] {
  for (let i = 0; i < documents.length; i++) {
    const document = documents[i]
    try {
      const node = parseDocument(document, srcDir)
      document.errors = validateGraphQlDocuments(schema, [
        node,
      ]) as GraphQLError[]

      const operation = node.definitions.find(
        (v) => v.kind === 'OperationDefinition',
      ) as OperationDefinitionNode | undefined
      if (operation) {
        document.name = operation.name?.value
        document.operation = operation.operation
      }

      // document.name = node
      document.isValid = document.errors.length === 0
    } catch (e) {
      document.errors = [e as GraphQLError]
      document.isValid = false
    }
  }

  return documents
}

/**
 * Generates the TypeScript definitions and documents files.
 */
export async function generate(
  options: ModuleOptions,
  schemaPath: string,
  resolver: Resolver['resolve'],
  srcDir: string,
  logEverything = false,
) {
  const schemaContent = await fsp.readFile(schemaPath).then((v) => v.toString())
  const schema = await loadSchema(schemaContent, { loaders: [] })

  const documents = await buildDocuments(
    options.documents,
    options.autoImportPatterns as string[],
    resolver,
  )

  const validated = validateDocuments(schema, documents, srcDir)

  const templates = await generateTemplates(
    validated.filter((v) => v.isValid).map((v) => v.content),
    schemaPath,
    options,
  )

  const hasErrors = validated.some((v) => !v.isValid)
  if (hasErrors || logEverything) {
    const table = new Table({
      head: ['Operation', 'Name', 'File', 'Errors'].map((v) => chalk.white(v)),
    })

    validated.forEach((document) => {
      if (logEverything || !document.isValid) {
        table.push(
          [
            document.operation || '',
            document.name || '',
            document.filename?.replace(srcDir, '') || '',
            document.errors?.join('\n\n') || '',
          ].map((v) => {
            if (document.isValid) {
              return v
            }
            return chalk.red(v)
          }),
        )
      }
    })

    logger.log(table.toString())
  }

  logger.info('Finished GraphQL code generation.')

  return {
    templates: templates.sort((a, b) => {
      return a.filename.localeCompare(b.filename)
    }),
    hasErrors,
    documents: validated.sort((a, b) => {
      if (a.filename && b.filename) {
        return a.filename.localeCompare(b.filename)
      }
      return -1
    }),
  }
}

export const fileExists = (
  path?: string,
  extensions = ['js', 'ts'],
): string | null => {
  if (!path) {
    return null
  } else if (existsSync(path)) {
    // If path already contains/forces the extension
    return path
  }

  const extension = extensions.find((extension) =>
    existsSync(`${path}.${extension}`),
  )

  return extension ? `${path}.${extension}` : null
}

export async function outputDocuments(
  outputPath: string,
  documents: GraphqlMiddlewareDocument[],
) {
  await fsp.mkdir(outputPath, { recursive: true })
  documents.forEach((v) => {
    if (v.operation && v.name) {
      const fileName = [v.operation, v.name, 'graphql'].join('.')
      const filePath = resolve(outputPath, fileName)
      fsp.writeFile(filePath, v.content)
    }
  })
}
