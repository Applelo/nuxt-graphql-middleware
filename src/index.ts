import path from 'path'
import fs from 'fs'
import mkdirp from 'mkdirp'
import chokidar from 'chokidar'
import { Context, Module } from '@nuxt/types'
import consola from 'consola'
import serverMiddleware, {
  GraphqlServerMiddlewareConfig,
} from './serverMiddleware'
import graphqlImport from './graphqlImport'
import { GraphqlMiddlewarePluginConfig } from './plugin'

const logger = consola.withTag('nuxt-graphql-middleware')

const PLUGIN_PATH = path.resolve(__dirname, 'plugin.js')

export interface GraphqlMiddlewareConfig {
  graphqlServer: string
  endpointNamespace?: string
  debug: boolean
  queries: Record<string, string>
  mutations: Record<string, string>
  outputPath: string
  plugin?: GraphqlMiddlewarePluginConfig
  server?: GraphqlServerMiddlewareConfig
}

enum FileType {
  Query = 'query',
  Mutation = 'mutation',
}

interface FileMapItem {
  type: FileType
  name: string
  file: string
}

function resolveGraphqlFile(file: string, resolver: any): Promise<string> {
  return fs.promises
    .readFile(file)
    .then((buffer) => buffer.toString())
    .then((source) => graphqlImport(source, resolver))
}

function writeSource(dest: string, filePath: string, source: string) {
  const fileName = path.basename(filePath)
  const out = path.resolve(dest, fileName)
  return fs.promises.writeFile(out, source)
}

function resolveGraphql(
  files: Record<string, string>,
  map: Map<string, string>,
  resolver: any,
  filesMap: Map<string, FileMapItem>,
  type: FileType,
  outputPath: string
) {
  return Promise.all(
    Object.keys(files).map((name) => {
      const filePath = files[name]
      const file = resolver(filePath)
      return resolveGraphqlFile(file, resolver).then((source) => {
        map.set(name, source)
        if (outputPath) {
          writeSource(outputPath, filePath, source)
        }
        filesMap.set(file, {
          type,
          name,
          file,
        })
      })
    })
  )
}

/*
 * Install the Nuxt GraphQL Middleware module.
 */
const graphqlMiddleware: Module = async function () {
  const options = this.options
  const provided = (this.options.graphqlMiddleware ||
    {}) as Partial<GraphqlMiddlewareConfig>

  const config: GraphqlMiddlewareConfig = {
    graphqlServer: provided.graphqlServer || '',
    endpointNamespace: provided.endpointNamespace || '/__graphql_middleware',
    debug: provided.debug || options.dev,
    queries: provided.queries || {},
    mutations: provided.mutations || {},
    outputPath: provided.outputPath || '',
    server: provided.server,
    plugin: {
      enabled: !!provided.plugin?.enabled,
      cacheInBrowser: !!provided.plugin?.cacheInBrowser,
      cacheInServer: !!provided.plugin?.cacheInServer,
    },
  }
  const resolver = this.nuxt.resolver.resolvePath

  // Add the API helper plugin.
  if (config.plugin?.enabled) {
    this.addPlugin({
      src: PLUGIN_PATH,
      options: {
        namespace: config.endpointNamespace,
        cacheInBrowser: config.plugin?.cacheInBrowser ? 'true' : 'false',
        cacheInServer: config.plugin?.cacheInServer ? 'true' : 'false',
      },
    })
  }

  const fileMap: Map<string, FileMapItem> = new Map()
  const queries = new Map()
  const mutations = new Map()

  const outputPath = config.outputPath ? resolver(config.outputPath) : ''
  if (outputPath) {
    await mkdirp(outputPath)
  }

  /**
   * Build all queries and mutations.
   */
  function build() {
    logger.log('Building GraphQL files...')
    return Promise.all([
      resolveGraphql(
        config.queries,
        queries,
        resolver,
        fileMap,
        FileType.Query,
        outputPath
      ),
      resolveGraphql(
        config.mutations,
        mutations,
        resolver,
        fileMap,
        FileType.Mutation,
        outputPath
      ),
    ]).then(() => {
      logger.success('Finished building GraphQL files')
    })
  }

  /**
   * Watch *.graphql files and rebuild everything on change.
   */
  function watchFiles() {
    const ignored = ['node_modules', '.nuxt']
    if (config.outputPath) {
      ignored.push(config.outputPath)
    }
    const filesWatcher = (options._filesWatcher = chokidar.watch(
      './**/*.graphql',
      {
        ignoreInitial: true,
        ignored,
      }
    ))

    if (filesWatcher) {
      logger.info('Watching for query changes')
      filesWatcher.on('change', () => {
        build()
      })
    }
  }

  this.nuxt.hook('build:done', () => {
    build().then(() => {
      if (options.debug) {
        logger.info('Available queries and mutations:')
        console.table(
          Array.from(fileMap.entries()).map(([_key, value]) => value)
        )
      }
    })
    if (this.nuxt.options.dev) {
      watchFiles()
    }
  })

  // Add out server middleware to manage the cache.
  this.addServerMiddleware({
    path: config.endpointNamespace,
    handler: serverMiddleware(
      config.graphqlServer,
      queries,
      mutations,
      config.server
    ),
  })
}

export default graphqlMiddleware
