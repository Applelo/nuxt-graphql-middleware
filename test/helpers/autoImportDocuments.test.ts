import path from 'path'
import { describe, expect, test } from 'vitest'
import { autoImportDocuments } from '../../src/helpers'

describe('autoImportDocuments', () => {
  const srcDir = path.resolve(__dirname, './../../playground')
  const resolver = () => srcDir

  test('Returns an empty array if no pattern is provided.', async () => {
    expect(await autoImportDocuments([], resolver)).toEqual([])
  })

  test('Returns all available document contents.', async () => {
    expect(await autoImportDocuments([], resolver)).toMatchSnapshot()
    const result = await autoImportDocuments(['**/*.graphql'], resolver)
    const data = JSON.stringify(result)
    expect(data).not.toContain('schema.graphql')
    expect(data).not.toContain('schema.gql')
  })

  test('Does not include the GraphQL schema file.', async () => {
    const result = await autoImportDocuments(['**/*.graphql'], resolver)
    const data = JSON.stringify(result)
    expect(data).not.toContain('schema.graphql')
    expect(data).not.toContain('schema.gql')
  })
})
