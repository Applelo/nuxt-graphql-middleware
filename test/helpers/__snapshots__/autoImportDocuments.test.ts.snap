// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`autoImportDocuments > Returns all available document contents. 1`] = `[]`;
